### PR TITLE
fix(connectors): reject null data values in vault.kv.patch schema

### LINKED ISSUE #1435: Task: G3.15 follow-up — vault.kv.patch data schema permits null (JSON-merge key-delete vs add/overwrite docs)
Parent goal: #214
Parent initiative: #1399

## Problem

`vault.kv.patch` (added in #1409 / PR #1426) documents "add/overwrite-only" semantics, but its `data` parameter schema permits `null` values. Under HashiCorp Vault's JSON Merge Patch (`secrets.kv.v2.patch`), a `null` value **deletes** the key rather than setting it — so a caller can silently delete a secret field through an op documented as additive. Flagged as a non-blocking Major by CodeRabbit during the #1426 review; the op is approval-gated so it's low-severity, but the schema/docs contract is internally inconsistent.

## Current state

- `vault.kv.patch` registered in `backend/src/meho_backplane/connectors/vault/ops.py`; its parameter schema allows `null` within `data`.
- Handler delegates to hvac `secrets.kv.v2.patch` (JSON Merge Patch semantics).
- Docstrings / `docs/codebase/connectors-vault.md` describe patch as merge/overwrite.

## Desired state

The schema and the documented semantics agree. Choose one:
1. **Reject `null`** in the `data` schema (`additionalProperties`/value typing that disallows null) so the op is genuinely add/overwrite-only — matches the docs.
2. **OR** document the JSON-Merge-Patch key-delete behavior explicitly (update docstring + connectors-vault.md) and keep `null` allowed as an intentional delete-a-field capability.

Option 1 is the lower-surprise default unless field-deletion-via-patch is a wanted capability.

## Acceptance criteria

- [ ] Schema and documentation are consistent about whether `null` is allowed and what it does.
- [ ] If `null` is rejected: a unit test asserts the schema rejects a `null` data value.
- [ ] If `null` is documented as delete: docstring + connectors-vault.md state it, with a test asserting the delete-via-null behavior.

## Out of scope

- Other vault ops; the approval-gating posture (unchanged).

## References

- Origin: #1409 / PR #1426 (G3.15-T1); CodeRabbit M1 on that PR.
- `backend/src/meho_backplane/connectors/vault/ops.py` (kv.patch schema + handler).

### PR DESCRIPTION
**Pull request by damir-topic**
## Summary

- `vault.kv.patch` (G3.15-T1 #1409) delegates to hvac's `secrets.kv.v2.patch`, which speaks JSON Merge Patch (RFC 7396): a `null` value **deletes** the key. The op is documented add/overwrite-only, but its `data` schema permitted `null` — letting a caller silently delete a secret field through an op that never advertises deletion (CodeRabbit M1 on #1426).
- Resolved per the issue's preferred **Option 1 (reject `null`)**: the `data` `additionalProperties` subschema now pins each value to a non-null JSON type (`string`/`number`/`boolean`/`object`/`array`), so a `null` fails validation as `invalid_params` before reaching Vault. Field removal stays explicit via `kv.put` (wholesale replace) or `kv.delete` (version soft-delete).
- Schema, handler docstring, `llm_instructions`, and `docs/codebase/connectors-vault.md` now all agree the op is add/overwrite-only.

Chose Option 1 because the op's docs were already uniformly additive, KV-v2 already exposes explicit deletion paths (`kv.put`/`kv.delete`), and the issue names Option 1 the lower-surprise default absent a wanted delete-via-patch capability — none exists here.

## Test plan

- [x] `ruff check` (ops.py, test file) — clean
- [x] `ruff format --check` — clean
- [x] `mypy src/.../vault/ops.py` — clean
- [x] `pytest tests/test_connectors_vault.py` — 64 passed (added 2 dispatcher param cases + 1 schema-level test)
- [x] `pytest tests/test_broadcast_events.py` — 83 passed (no regression)
- [x] `code-quality.py --diff` — 0 blockers
- [x] Acceptance: schema + docs consistent (null rejected, add/overwrite-only); unit test `test_vault_kv_patch_schema_rejects_null_data_value` asserts the schema rejects a `null` data value
- [x] CLI snapshot regen (`make snapshot-openapi && make generate`) — ran; no change to `cli/api/openapi.json` (typed-op param schemas live in `endpoint_descriptor`/`describe_operation`, not the static OpenAPI doc)

## Out of scope

- Other vault ops; the approval-gating posture (unchanged per the issue).

Closes #1435

### CODE REVIEW

**Comment by damir-topic** [comment]
@coderabbitai review

File: backend/src/meho_backplane/connectors/vault/ops.py [hunk 1]
```diff
@@ -521,11 +534,17 @@ async def vault_kv_put(operator: Operator, target: Any, params: dict[str, Any])
         "data": {
             "type": "object",
             "minProperties": 1,
+            "additionalProperties": {
```
Comment (damir-topic) [inline backend/src/meho_backplane/connectors/vault/ops.py]: **[🛑 B1]** — null-rejection only covers top-level `data` values; nested nulls still delete keys in Vault

**What:** This `additionalProperties` subschema constrains the *direct* values of `data` to non-null types, but a nested `null` — e.g. `{"data": {"creds": {"old": null}}}` — has a nested-object value of type `object`, which validates, and the inner `null` is never checked. Verified against `Draft202012Validator(VAULT_KV_PATCH_PARAMETER_SCHEMA)`: top-level `{"k": null}` is rejected, nested `{"obj": {"inner": null}}` is accepted.

**Why it matters:** hvac's `secrets.kv.v2.patch` speaks JSON Merge Patch (RFC 7396), whose algorithm recurses — `MergePatch(Target[Name], Value)` — so a `null` at *any* depth deletes that key (RFC 7396 example: `{"a": {"b": "d", "c": null}}` removes nested `c`). The docstring, `llm_instructions`, and `connectors-vault.md` all now claim this op pins *each* `data` value to a non-null type and is *genuinely* add/overwrite-only. For nested objects that claim is false: a nested `null` reaches Vault and silently deletes the nested field — the exact schema-vs-docs inconsistency #1435 was filed to eliminate, relocated to depth >= 1.

**Required change:** Make schema and docs agree. Either (a) recurse the non-null constraint into nested objects/arrays (e.g. a referenced "non-null JSON value" subschema applied via `additionalProperties` on nested objects and `items` on arrays), or (b) scope the docstring + `llm_instructions` + `docs/codebase/connectors-vault.md` claim explicitly to *top-level* `data` values and state that nested object merges still follow RFC 7396 delete-on-null. Add a test covering the nested case in whichever direction is chosen.

**Source:** RFC 7396 section 1 algorithm (recursive MergePatch), https://www.rfc-editor.org/rfc/rfc7396 ; Vault KV-v2 PATCH uses application/merge-patch+json, https://developer.hashicorp.com/vault/api-docs/secret/kv/kv-v2

**Review by damir-topic — COMMENTED** [review]
## /auto-review-pr summary (iteration 1)




**Anchor task:** #1435 — vault.kv.patch `data` schema permits `null` (JSON-merge key-delete vs add/overwrite docs)
**Surface area:** backend
**Best-practices packs audited:** python
**Verdict:** REQUEST_CHANGES
**Head SHA reviewed:** `19ce7591`
**AI reviewer (CodeRabbit):** not received within window — CodeRabbit returned an explicit **rate-limit** response ("review limit reached… available in 57 minutes"; org out of usage credits), not a review. No findings to cross-check this iteration. (Polled 183s; full 15-min wait would be futile against a deterministic rate-limit block.)

### Acceptance criteria
- [x] Schema rejects a top-level `null` data value — verified: `Draft202012Validator(VAULT_KV_PATCH_PARAMETER_SCHEMA)` rejects `{"data": {"k": null}}`; dispatcher returns `invalid_params`. Unit test `test_vault_kv_patch_schema_rejects_null_data_value` + two dispatcher param cases (`rejects-null-value`, `rejects-null-among-non-null`) assert it.
- [❌] Schema and documentation are **consistent** about whether `null` is allowed — **not fully met.** The docstring / `llm_instructions` / `docs/codebase/connectors-vault.md` assert an absolute guarantee ("pins **each** `data` value to a non-null type", "**genuinely** add/overwrite-only"), but the `additionalProperties` subschema only constrains **top-level** values. A nested `null` (e.g. `{"data": {"creds": {"old": null}}}`) passes validation and — per RFC 7396, which recurses — Vault's JSON Merge Patch deletes the nested `old` key. The schema-vs-docs inconsistency #1435 set out to eliminate is reintroduced at nesting depth ≥ 1. See B1.

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #1435 | Open (task, priority:low) | Closes | ✅ | ⚠️ partial (top-level only) | close on merge once B1 addressed |

Parent initiative #1399 / parent goal #214 confirmed from issue body. No sibling open PR touches `vault/ops.py` (#1460 ci, #1461 testcontainers, #1463 composite preflight) — no collision.

### Predicted vs. actual diff
- ✅ `backend/src/meho_backplane/connectors/vault/ops.py` — schema `additionalProperties` value-type constraint + docstring/`llm_instructions` updates
- ✅ `backend/tests/test_connectors_vault.py` — schema-level test + two dispatcher param cases
- ✅ `docs/codebase/connectors-vault.md` — add/overwrite-only contract note
- 🚫 none undisclosed
- 🕳️ none missing

### Integration verification
- [x] `ruff check` (ops.py, test file) — clean
- [x] `ruff format --check` — clean
- [x] `mypy src/.../vault/ops.py` — clean
- [x] `pytest tests/test_connectors_vault.py` — 64 passed (incl. 3 new null-rejection assertions), run with local-src `PYTHONPATH` (editable install pointed at a different worktree)
- [x] Schema behavior probed directly: top-level `null` rejected ✅; nested `null` accepted ⚠️ (see B1)
- [⚠️ not run] full backend suite / migrations / docker — out of scope for this slice

### Findings (inline — each is a merge blocker per team policy)
For the `/auto-implement-issue --continue-pr` punch list:

- 🛑 B1 — `data` null-rejection only covers top-level values; nested `null` still reaches Vault and deletes the nested key, contradicting the PR's "each value / genuinely add/overwrite-only" docs. Required change: either recurse the non-null constraint into nested objects/arrays in the schema, OR scope the docstring + `connectors-vault.md` + `llm_instructions` claim to **top-level** values (and note nested-object merges still follow RFC 7396 delete-on-null), so schema and docs agree. Add a test for the nested case in whichever direction is chosen.

### Adjacent improvements (non-blocking, not posted inline)
- The `_VAULT_KV_PATCH_RESPONSE_SCHEMA` allows `version: null` while the handler always returns an int from `patch_payload["data"]["version"]`; harmless, pre-existing, out of scope here.

### AI reviewer (CodeRabbit) cross-check
- Adopted: 0
- Disputed: 0
- Out of scope: 0
- CodeRabbit did not run (rate-limited / org out of credits). B1 is an independent reviewer finding.

### Docs
- [x] `docs/codebase/connectors-vault.md` — updated by the PR, but the update overstates the guarantee (see B1). `docs_status: updated` (with a consistency defect).

### What worked well (review-body only)
- Clean, minimal, idiomatic JSON Schema 2020-12 fix (`type` array excluding `"null"`) validated by the same `Draft202012Validator` the dispatcher uses.
- Schema-level test asserts the contract independently of dispatcher wiring — good defense against future refactors.
- Docstring/`llm_instructions`/docs all updated in lockstep with a clear RFC 7396 rationale.

**Comment by damir-topic** [comment]
@coderabbitai review

**Review by damir-topic — COMMENTED** [review]
## /auto-review-pr summary (iteration 2)




**Anchor task:** #1435 — vault.kv.patch `data` schema permits null (JSON-merge key-delete vs add/overwrite docs)
**Surface area:** backend
**Best-practices packs audited:** python
**Verdict:** APPROVE
**Head SHA reviewed:** `c088adbe8c222dd0daa7eb4bb592ea2764ef6a72`
**AI reviewer (CodeRabbit):** not received within 15min (re-triggered at the fix push; org still rate-limited from the iter-1 back-to-back run — known pattern, does not block this verdict)

This is a re-review after the iter-1 fix for blocker **B1**.

### B1 — RESOLVED (verified empirically)

The fix replaced the flat per-value type constraint with a recursive
non-null subschema `$defs/nonNullJsonValue` (scalar | object whose values
`$ref` back | array whose items `$ref` back), applied via
`additionalProperties` on objects and `items` on arrays. I verified the
same-document `#/$defs/nonNullJsonValue` `$ref` actually resolves under the
dispatcher's `Draft202012Validator(parameter_schema)` (the exact construction
in `operations/_validate.py:87`) by importing the real
`VAULT_KV_PATCH_PARAMETER_SCHEMA` and probing:

| Shape | Expected | Result |
|---|---|---|
| `{"data":{"a":null}}` | reject | ✅ rejected |
| `{"data":{"a":{"b":null}}}` | reject | ✅ rejected |
| `{"data":{"a":[null]}}` | reject | ✅ rejected |
| `{"data":{"a":[{"b":null}]}}` | reject | ✅ rejected |
| `{"data":{"a":{"b":[{"c":null}]}}}` | reject | ✅ rejected |
| `{"data":{"a":"x"}}` / number / bool | accept | ✅ valid |
| `{"data":{"a":{"b":"x"}}}` | accept | ✅ valid |
| `{"data":{"a":[1,"x"]}}` | accept | ✅ valid |
| `{"data":{"a":[{"b":"x"}]}}` | accept | ✅ valid |
| `{"data":{"a":{"b":[{"c":"x"}]}}}` | accept | ✅ valid |

`Draft202012Validator.check_schema` passes; the `oneOf` branches are
mutually exclusive by `type` (no `oneOf` ambiguity); empty nested
`object`/`array` still validate (no over-restriction); integers, floats,
negatives, and zero validate via the `number` branch. A null at any depth
fails, equivalent non-null structures pass. The schema-vs-docs contradiction
#1435 was filed to close no longer exists at any nesting depth.

### Acceptance criteria
- [x] Nested null (`{"data":{"creds":{"old":null}}}`) rejected before reaching Vault — verified empirically + by new tests
- [x] Docs / docstring / `llm_instructions` agree with the schema ("add/overwrite-only at every depth") — read all three, consistent
- [x] Tests cover the nested case — `rejects-nested-object-null` / `rejects-nested-array-null` (dispatcher) + `test_vault_kv_patch_schema_rejects_nested_null_data_value` (schema-level, with non-null acceptance assertions)

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #1435 | Open (task) | Closes | ✅ | ✅ full | close on merge |

`closingIssuesReferences` confirms #1462 closes #1435. Parent initiative #1399 / goal #214. No sibling open PR touches `vault/ops.py`.

### Predicted vs. actual diff
- ✅ `backend/src/meho_backplane/connectors/vault/ops.py` — recursion subschema + doc text only; scoped to B1
- ✅ `backend/tests/test_connectors_vault.py` — nested-null cases added at both levels
- ✅ `docs/codebase/connectors-vault.md` — "add/overwrite-only at every depth" prose matches code
- 🚫 none undisclosed · 🕳️ none missing

### Integration verification
- [x] `ruff check` — clean (ops.py + test file)
- [x] `ruff format --check` — 2 files already formatted
- [x] `mypy` — no issues in ops.py
- [x] `pytest tests/test_connectors_vault.py` — 67 passed (was 64 at iter-1; +3 net), incl. the new nested-null cases
- [x] empirical `Draft202012Validator` probe — 13 shapes, all as expected
- [x] CI (GitHub Actions): Python ruff+mypy+pytest, integration testcontainers, Semgrep SAST, CLI API snapshot freshness, Helm, Go — all SUCCESS
- [⚠️ not run] live Vault round-trip — covered by the dev-e2e suite in CI, not re-run locally

The two ops.py/test files in the review worktree are byte-identical (sha256 `be17e9e…`) to the editable-install path pytest resolved, so the 67-pass run reflects the PR code under review.

### Findings
None blocking. B1 cleared.

### Adjacent improvements (non-blocking, not posted inline)
- The recursive `$ref` schema descends one validation frame per nesting level; at Python's default recursion limit a payload nested deeper than ~150 levels would raise `RecursionError` rather than a clean `invalid_params`. This is inherent to any recursive JSON Schema, far beyond any realistic Vault secret (single-digit depth), and the dispatcher's generic `except Exception` catches it — recorded for awareness only, not a required change.
- Pre-existing: `_VAULT_KV_PATCH_RESPONSE_SCHEMA` permits `version: null` while the handler always returns an int. Out of scope (noted in iter-1).

### AI reviewer (CodeRabbit) cross-check
- Adopted: 0 · Disputed: 0 · Out of scope: 0 — CodeRabbit did not return a review within the 15-min window (rate-limited, org out of credits). B1 was an independent reviewer finding; this APPROVE stands on the empirical verification above.

### Docs
- [x] `docs/codebase/connectors-vault.md` — updated, verified consistent with schema

### What worked well
- The recursion approach (option a) is the lower-surprise fix: it makes the schema match the already-shipped "add/overwrite-only" docs rather than weakening the docs to a top-level-only caveat.
- Tests assert both rejection (null at depth) and acceptance (equivalent non-null structures), so the guard can't silently over-restrict in a future regression.
